src/rgw/rgw_pubsub_push.cc

```diff
@@ -679,13 +684,115 @@ class RGWPubSubKafkaEndpoint : public RGWPubSubEndpoint {
 static const std::string KAFKA_SCHEMA("kafka");
 #endif	// ifdef WITH_RADOSGW_KAFKA_ENDPOINT
 
+#ifdef WITH_RADOSGW_AWS_ENDPOINT
+class RGWPubSubAWSEndpoint : public RGWPubSubEndpoint {
+private:
+
+  CephContext* const cct;
+  const std::string arn;
+  rgw::aws::AwsClient client;
+
+  bool get_verify_ssl(const RGWHTTPArgs& args) {
+    bool exists;
+    auto str_verify_ssl = args.get("verify-ssl", &exists);
+    if (!exists) {
+      // verify server certificate by default
+      return true;
+    }
+    boost::algorithm::to_lower(str_verify_ssl);
+    if (str_verify_ssl == "true") {
+      return true;
+    }
+    if (str_verify_ssl == "false") {
+      return false;
+    }
+    throw configuration_error("'verify-ssl' must be true/false, not: " + str_verify_ssl);
+  }
+
+  bool get_use_ssl(const RGWHTTPArgs& args) {
+    bool exists;
+    auto str_use_ssl = args.get("use-ssl", &exists);
+    if (!exists) {
+      // by default ssl not used
+      return false;
+    }
+    boost::algorithm::to_lower(str_use_ssl);
+    if (str_use_ssl == "true") {
+      return true;
+    }
+    if (str_use_ssl == "false") {
+      return false;
+    }
+    throw configuration_error("'use-ssl' must be true/false, not: " + str_use_ssl);
+  }
+
+  std::string get_arn(const RGWHTTPArgs& args){
+    bool exists;
+    auto arnString = args.get("AwsArn", &exists);
+    boost::optional<ARN> arn1 = ARN::parse(arnString);
+    if(!arn1) {
+      throw configuration_error("incorrect arn format");
+    }
+    if(arn1->service != Service::sns && arn1->service != Service::lambda){
+      throw configuration_error("only sns and lambda aws services allowed");
+    }
+    return arn1->to_string();
+  }
+
+  std::string get_access_key(const RGWHTTPArgs& args){
+    bool exists;
+    auto accessKey = args.get("AccessKey", &exists);
+    if(!exists) {
+      return "";
+    }
+    return accessKey;
+  }
+
+  std::string get_access_secret(const RGWHTTPArgs& args){
+    bool exists;
+    auto accessSecret = args.get("SecretKey", &exists);
+    if(!exists) {
+      return "";
+    }
+    return accessSecret;
+  }
+
+public:
+  RGWPubSubAWSEndpoint(const std::string& _endpoint,
+      const RGWHTTPArgs& args,
+      CephContext* _cct) :
+        cct(_cct),
+        arn(get_arn(args)),
+        client(rgw::aws::connect(get_access_key(args), get_access_secret(args), arn, "", true)){
+    if (client == rgw::aws::AwsClient()) {
+      throw configuration_error("AWS: failed to create connection to: " + _endpoint);
+    }
+  }
+
+  RGWCoroutine* send_to_completion_async(const rgw_pubsub_s3_record& record, RGWDataSyncEnv* env) override {}
```

Comment:
return `nullptr` to avoid compilation issue
you can also add a `ceph_assert()` here to make sure these are not called

Reply:
how to use `ceph_assert()` ? please explain this a bit

Reply:
```cpp
#include "include/ceph_assert.h"
...
ceph_assert(false); // or any boolean conditions
```